api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,302 @@
+// containing either the next chunk of the request body available, or an end-of-stream.
+// Once this is returned, it will no longer be possible to modify the
+// message headers.
+message SendRequestBodyChunkCommand {
+}
+
+// Send the next chunk of the response body.
+message SendResponseBodyChunkCommand {
+}
+
+// Request that Envoy will buffer the entire request message body in memory
+// subject to the currently-configured body buffer limit, and send a
+// single Body message downstream containing the entire message body.
+// It is still possible for the server to modify the message headers
+// with the next upstream message.
+message SendBufferedRequestBodyCommand {
+}
+
+// Request that Envoy buffer and send the response body in a single chunk.
+message SendBufferedResponseBodyCommand {
+}
+
+// Request that Envoy will send the HTTP request trailers when they are
+// available. Once trailers have been requested, no more bodies
+// will be delivered, so this must be sent after all desired body chunks are
+// desired.
+message SendRequestTrailersCommand {
+}
+
+// Request that Envoy send the response trailers when available.
+message SendResponseTrailersCommand {
+}
+
+// The mutation server may respond with zero or more of these responses
+// before closing the stream.
+// [#next-free-field: 16]
+message MutationResponse {
+  oneof command {
+    option (validate.required) = true;
+
+    StopCommand stop = 1;
+
+    ContinueRequestCommand continue_request = 2;
+
+    ContinueResponseCommand continue_response = 3;
+
+    SendRequestBodyChunkCommand send_request_body_chunk = 4;
+
+    SendResponseBodyChunkCommand send_response_body_chunk = 5;
+
+    SendBufferedRequestBodyCommand send_buffered_request_body = 6;
+
+    SendBufferedResponseBodyCommand send_buffered_response_body = 7;
+
+    SendRequestTrailersCommand send_request_trailers = 8;
+
+    SendResponseTrailersCommand send_response_trailers = 9;
+  }
+
+  // Replace the entire HTTP response (whether on the request or on the
+  // response path) with the contents of the CompleteResponse message.
+  // If a StopCommand command is set, then send this response back immediately
+  // to the client with no additional filter processing.
+  CompleteResponse complete_response = 10;
+
+  // Add or replace HTTP headers. This will be ignored after SendXXXBodyChunkCommand has
+  // been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated config.core.v3.HeaderValueOption set_headers = 11;
+
+  // Remove these HTTP headers. This will be ignored after SendXXXBodyChunk
+  // has been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated string remove_headers = 12;
+
+  // Replace some or all of the last message body received with a new
+  // set of bytes. This will be ignored unless the previous message sent
+  // by the client is a "Body" request.
+  BodyMutation body_mutation = 13;
```

Comment:
It's not clear from an API perspective where these apply. I think all of these should be part of relevant commands above?

Reply:
This has gotten confusing (same as for the "oneof" conversation above). This is basically a request / response protocol on top of a gRPC stream. I'm going to try and reorganize this proto to that effect and see it it helps make it easier for an external server developer to serve this protocol properly without breaking the proxy.